sdk/eventhub/Azure.Messaging.EventHubs/src/EventProcessorClient.cs

```diff
@@ -287,194 +437,404 @@ public class EventProcessorClient : EventProcessorBase<PartitionContext>, IAsync
+        ///
+        public virtual void StartProcessing(CancellationToken cancellationToken = default) => StartProcessingAsync(cancellationToken).GetAwaiter().GetResult();
+
+        /// <summary>
+        ///   Stops the event processor.  In case it isn't running, nothing happens.
         /// </summary>
         ///
-        /// <param name="reason">The reason why the processing for the associated partition is being stopped.</param>
-        /// <param name="context">The context in which the associated partition was being processed.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
         ///
         /// <returns>A task to be resolved on when the operation has completed.</returns>
         ///
-        protected async override ValueTask ProcessingForPartitionStoppedAsync(ProcessingStoppedReason reason,
-                                                                              PartitionContext context)
+        public virtual async Task StopProcessingAsync(CancellationToken cancellationToken = default)
         {
-            if (ProcessingForPartitionStoppedAsyncHandler != null)
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+            if (ActiveLoadBalancingTask != null)
             {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
                 try
                 {
-                    var stopContext = new PartitionProcessingStoppedContext(context, reason);
-                    await ProcessingForPartitionStoppedAsyncHandler(stopContext);
+                    cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                    if (ActiveLoadBalancingTask != null)
+                    {
+                        // Cancel the current running task.
+
+                        RunningTaskTokenSource.Cancel();
+                        RunningTaskTokenSource = null;
+
+                        // Now that a cancellation request has been issued, wait for the running task to finish.  In case something
+                        // unexpected happened and it stopped working midway, this is the moment we expect to catch an exception.
+
+                        try
+                        {
+                            await ActiveLoadBalancingTask.ConfigureAwait(false);
+                        }
+                        catch (Exception ex) when (ex is TaskCanceledException || ex is OperationCanceledException)
+                        {
+                            // Nothing to do here.  These exceptions are expected.
+                        }
+                        catch (Exception)
+                        {
+                            // TODO: delegate the exception handling to an Exception Callback.  Instead of delegating it to the handler,
```

Comment:
Yes... we should log (eventually) and bubble the exception via `throw`.    We have a deterministic point where callers can observe the exception.   The reason the handler exists is that there's otherwise no place to signal that something broked.

Reply:
I tried surfacing it, but we are getting some weird exceptions in some tests (some of them are `NullPointerException`s iirc).

Given that the sample still works and processes partitions correctly, we're probably just doing something silly when stopping the processor (a race condition maybe?). I hid the exception again to unblock merging, but I'll take a deeper look on that next week.

Reply:
Works for me.   That's something we can definitely tune after code complete (or release)